### ISSUE TITLE
training error on new dataset

### ISSUE THREAD

**Issue by xiaoli1996** [RicherMans/CED#18]
Sorry to bother you again. I'm training a new dataset with 7-classes multi-label. Follow your example, my "train.tsv" and "test.tsv" like this :
filename	labels	hdf5path
/XXXX.wav	2;0;5	data/hdf5/test_0.h5

then run the "run.py" python run.py train trainconfig/balanced_mixup_base_T_ensemble5014_chunk10.yaml --logitspath ./logits/audiotransformer_base_4740/train_sleepaudio/chunk_10/ --train_data ./data/labels/train.tsv --eval_data ./data/labels/test.tsv --num_classes 7 

![image](https://github.com/RicherMans/CED/assets/52903099/b6198778-401a-4114-b78b-beab71bdc69d)

print "idx"  is ranged 0 to 506, the classes of Audioset are 527, but my dataset is 7 classes.  What should I do to fix it?

**Comment by RicherMans** [comment]
Hey there,
So the error is due to the evaluation set in Audioset having 527 labels. If you want to evaluate on other data, just pass that data accordingly.

However, it seems you misunderstood the purpose of CED.
The framework itself is `label-free`, which means a teacher (or an ensemble) predict logits (i.e., targets) for a student model.
Since your teacher classifies 527 labels, it means that you should train with 527 targets.

Since I quite don't get your usecase here, I might ask:
Did you try to fine-tune using a pretrained CED model? Was the performance bad or why would you need to -rerun CED on a small scale dataset?

Kind regards

**Comment by xiaoli1996** [comment]
Thank you for your timely reply. I aim to run private datasets to validate the robustness of the dataset using different audio classification models. These models are trained from scratch and fine-tuned with the pre-trained models, respectively.

**Comment by RicherMans** [comment]
Hey,
So I am still not sure what you mean exactly by " These models are trained from scratch and fine-tuned with the pre-trained models, respectively." 
So you want to compare training from scratch vs. using a pretrained model from Audioset?

**Comment by xiaoli1996** [comment]
Please excuse the phrasing. These models operate under two paradigms: 1. they are trained from scratch using a proprietary dataset, 2. fine-tuned from a pre-trained version of the same model.

**Comment by RicherMans** [comment]
Alright got you, so if that is the case, why do you need to use CED? I mean the training paradigm CED, not the models (CED-tiny/mini/small/base) ?

**Comment by xiaoli1996** [comment]
Maybe I don't fully understand the CED code, I want to try to train CED-base from scratch  and fine-tune by a pretrained CED-base on my dataset.

**Comment by RicherMans** [comment]
Yeah, that would be possible but in that case you would need Audioset for training.


If you would like to train a model with CED on your dataset you would need to:
1. Train some models on your dataset with num_classes = 7. Lets say you train 5 models.
2. Use these 5 models as teachers, where you can use `save_logits.py` to prepare for training.
3. Train a student model ( can be much smaller than teachers ) on these logits (from 5 teachers). Note that this student actually would use `num_classes=7`, since the teacher also has the same setting.

You would use CED if you have already trained a lot of models on your dataset, but would like to further improve performance.

Otherwise, just fine-tune from one of our checkpoints. In my experience they should always outperform pretty much everything else that is publicly available.

Kind regards,
Heinrich

**Comment by xiaoli1996** [comment]
Thank you very much! I will try your suggestions and ask for help if I have any questions. I will close this issue later.
